tools/tpm2_policypcr.c

```diff
@@ -0,0 +1,165 @@
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <ctype.h>
+
+#include <sapi/tpm20.h>
+
+#include "files.h"
+#include "log.h"
+#include "pcr.h"
+#include "tpm2_options.h"
+#include "tpm2_policy.h"
+#include "tpm2_session.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef struct tpm2_startauthsession_ctx tpm2_startauthsession_ctx;
+struct tpm2_startauthsession_ctx {
+   const char *session_path;
+   const char *raw_pcrs_file;
+   TPML_PCR_SELECTION pcr_selection;
+   const char *policy_out_path;
+};
+
+static tpm2_startauthsession_ctx ctx;
+
+static bool on_option(char key, char *value) {
+
+    switch (key) {
+    case 'f':
+        ctx.policy_out_path = value;
+        break;
+    case 'F':
+        ctx.raw_pcrs_file = value;
+        break;
+    case 'L': {
+        bool result = pcr_parse_selections(value, &ctx.pcr_selection);
+        if (!result) {
+            LOG_ERR("Could not parse PCR selections");
+            return false;
+        }
+    } break;
+    case 'S':
+        ctx.session_path = value;
+    break;
+    }
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static struct option topts[] = {
+        { "policy-file",    required_argument,  NULL,   'f' },
+        { "pcr-input-file", required_argument,  NULL,   'F' },
+        { "set-list",       required_argument,  NULL,   'L' },
+        { "session",        required_argument,  NULL,   'S' },
+    };
+
+    *opts = tpm2_options_new("f:F:L:S:", ARRAY_LEN(topts), topts, on_option,
+                             NULL, false);
+
+    return *opts != NULL;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+
+    UNUSED(flags);
+
+    int rc = 1;
+    tpm2_session *s = NULL;
+
+    bool fail = false;
+    if (!ctx.session_path) {
+        LOG_ERR("Must specify -S session file");
+        fail = true;
+    }
+
+    if (!ctx.pcr_selection.count) {
+        LOG_ERR("Must specify -L pcr selection list");
+        fail = true;
+    }
+
+    if (fail) {
```

Comment:
Not sure I understand the reason of this fail variable, why not just return 1 in the error path of the two checks above?

Reply:
Output all things wrong with arguments versus just one error at a time.